### ISSUE TITLE
Clojure Concurrency Tutorial for Beginners with Code Examples

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#3298]
Clojure Concurrency Tutorial for Beginners with Code Examples<br>
<br>
http://bit.ly/2uU70o3<br>
<br>
About 0 Minutes<br>
<br>
<div>
<div>
<p>Do you want to learn how to do concurrency in Clojure?</p>
<p>In this tutorial, you'll learn 16 ways for doing concurrency in Clojure including how to start threads and how to communicate between them.</p>
<h2>Table of Contents</h2>
<ul>
<li>
<a href="#catalog">Catalog of Primitives</a>
<ul>
<li>
<a href="#delay">Delay</a> - I may not need to calculate this</li>
<li>
<a href="#promise">Promise</a> - I'll check back here for the answer</li>
<li>
<a href="#future">Future</a> - Please calculate this in another thread</li>
<li>
<a href="#deref">A word about <code>deref</code></a> - What's in the box?</li>
<li>
<a href="#atom">Atom</a> - Keeping a single value consistent over time</li>
<li>
<a href="#ref">Ref</a> - Keep multiple values in a consistent relationship</li>
<li>
<a href="#var">Var</a> - Redefine stuff as you program</li>
<li>
<a href="#agent">Agent</a> - Queue up work to aggregate a value</li>
<li>
<a href="#lock">Lock</a> - Keep others out while you're in</li>
<li>
<a href="#core.async">core.async</a> - Workers with conveyor belts</li>
<li>
<a href="#threads">Threads</a> - Use your cores</li>
<li>
<a href="#executorservice">Executor Service</a> - Run lots of tasks in a thread pool</li>
<li>
<a href="#manifold">Manifold</a> - Abstractions for data sources</li>
<li>
<a href="#pmap"><code>pmap</code></a> - Simple parallelism with one more letter</li>
<li>
<a href="#reducers">Reducers</a> - Transformation pipelines in parallel</li>
<li>
<a href="#forkjoin">Fork/Join</a> - Break your work into chunks, do the chunks on your cores, put them back together</li>
</ul>
</li>
<li><a href="#secret">Concurrency vs parallelism</a></li>
<li><a href="#threadsjvm">JVM Thread model</a></li>
<li><a href="#ds">Immutable Data Structures</a></li>
<li><a href="#pure">Pure Functions</a></li>
<li><a href="#independence">Independence</a></li>
</ul>
<h2>Concurrency Primitive Catalog</h2>
<p>So let's go over a bunch of different primitives with examples.</p>
<h3>Delay</h3>
<p>Wouldn't it be great if you could write some code that calculates a value, but never run the code if no one wants the value? And wouldn't it also be great if the value was only calculated once, no matter how many threads need the value?</p>
<p>There's a Clojure function called <a href="https://clojuredocs.org/clojure.core/delay"><code>delay</code></a>. It creates an object called a Delay. A Delay ensures that some code is either run zero or one time. It's run zero times if the Delay is never <code>deref</code>ed. And it's run once if it is <code>deref</code>ed, regardless of how many times it's <code>deref</code>ed.</p>
<p>Here's an example:</p>
<pre>
<code><span>;; make a delay that does some calculation</span>
<span>(</span><span>def</span> the-answer <span>(</span>delay <span>(</span><span>*</span> <span>100</span> <span>(</span>long-calculation <span>45</span><span>)</span><span>)</span><span>)</span><span>)</span>

<span>;; that long calculation has not been done at this point</span>

<span>(</span><span>when</span> need-the-answer?
  <span>;; the thread will block until the answer is calculated</span>
  <span>(</span><span>println</span> <span>(</span><span>deref</span> the-answer<span>)</span><span>)</span><span>)</span>

<span>;; if you don't need the answer, it was never run</span></code>
</pre>
<p>Now, that's a nice way of avoiding work in a single thread if you're going to use the work some of the time but not all of the time. Delays are also useful for a shared resource between threads.</p>
<pre>
<code><span>;; make a delay that initializes a shared resource</span>
<span>;; if we don't put it in a Delay, the resource will be initialized during compilation</span>
<span>(</span><span>def</span> resource <span>(</span>delay <span>(</span>initialize-shared-resource<span>)</span><span>)</span><span>)</span>

<span>;; start 100 threads that use the resource</span>
<span>(</span><span>defn</span> -main <span>[</span><span>]</span>
  <span>(</span><span>dotimes</span> <span>[</span>x <span>100</span><span>]</span>
    <span>(</span><span>doto</span> <span>(</span>Thread. <span>(</span><span>fn</span> <span>[</span><span>]</span>
                     <span>(</span><span>let</span> <span>[</span>resource @resource<span>]</span>
                       <span>;; use the resource here ...</span>
                       <span>)</span><span>)</span>
      .start<span>)</span><span>)</span><span>)</span></code>
</pre>
<p>The magic is that only one of those threads will initialize it. The rest of the threads will block waiting for it.</p>
<p>Thanks to Dan Lebrero for <a href="http://danlebrero.com/2017/06/12/delay-clojure-forgotten-concurrency-primitive/">explaining how Delay is useful for concurrency</a>.</p>
<p>I've made a video lesson about this. Check out <a href="https://purelyfunctional.tv/lesson/delay/"><em>Delay</em></a>.</p>
<h3>Promise</h3>
<p>Clojure has a function called <a href="https://clojuredocs.org/clojure.core/promise"><code>promise</code></a>. A Promise is like a coupon for a back rub that you get from your significant other. It's not a present itself, but instead a Promise to give you a present later.</p>
<p>Calling <code>promise</code> will make a coupon for a value you promise to deliver some time in the future. Usually you are going to calculate that value in another thread.</p>
<p>Here's how you would use it.</p>
<pre>
<code><span>;; make a Promise</span>
<span>(</span><span>def</span> the-answer <span>(</span>promise<span>)</span><span>)</span>

<span>;; start working on the promise in a new thread</span>
<span>(</span><span>doto</span> <span>(</span>Thread. <span>(</span><span>fn</span> <span>[</span><span>]</span>
                 <span>;;; do a lot of work</span>
                 <span>;;; ...</span>
                 <span>;;; then deliver on your promise</span>
                 <span>(</span>deliver the-answer <span>42</span><span>)</span><span>)</span><span>)</span>
  .start<span>)</span>

<span>;; in the original thread, do your own work</span>
<span>;; ...</span>
<span>;; then get the answer</span>
<span>;; (this will block if it's not there)</span>
<span>(</span><span>println</span> <span>(</span><span>deref</span> the-answer<span>)</span><span>)</span></code>
</pre>
<p>Promises are a simple way to communicate between threads. One Thread calculates the answer. The other waits for it. (We'll look at how to create Threads later.)</p>
<p>You can create as many Promises as you want. But each one can only get one answer. Once the answer is delivered, it can't be delivered again.</p>
<p>I've made a video lesson about this. Check out <a href="https://purelyfunctional.tv/lesson/promise/"><em>Promise</em></a>.</p>
<h3>Future</h3>
<p>A Future is similar to a Promise. The main difference is that Futures will evaluate an expression for you in another Thread. You don't have to set the thread up. Here's how the same thing works using <a href="https://clojuredocs.org/clojure.core/future"><code>future</code></a>.</p>
<pre>
<code><span>(</span><span>def</span> the-answer <span>(</span>future
                   <span>;; do a lot of work in a new thread</span>
                   <span>;; ...</span>
                   <span>;; then deliver the answer</span>
                   <span>;; the value of the last expression is delivered</span>
                   <span>42</span><span>)</span><span>)</span>

<span>;; in the main thread, do some work</span>
<span>;; ...</span>
<span>;; then get the answer (which blocks until the answer is done)</span>
<span>(</span><span>println</span> <span>(</span><span>deref</span> the-answer<span>)</span><span>)</span></code>
</pre>
<p>One thing that trips up people new to Futures is that they swallow exceptions. If the code you run in the Future throws an exception, you won't hear about it until you <code>deref</code> it. When you <code>deref</code> it, the exception will be thrown again in the current thread.</p>
<p>Let's look at some code that demonstrates this:</p>
<pre>
<code><span>;; the Exception gets thrown but stored in the Future</span>
<span>(</span><span>def</span> <span>f</span> <span>(</span>future <span>(</span><span>throw</span> <span>(</span>Exception. <span>"Hello from the future!"</span><span>)</span><span>)</span><span>)</span><span>)</span>

<span>(</span><span>deref</span> <span>f</span><span>)</span> <span>;; this will throw the Exception</span></code>
</pre>
<p>I've made a video lesson about this. Check out <a href="https://purelyfunctional.tv/lesson/future/"><em>Future</em></a>.</p>
<h3>A word (or two) about <code>deref</code>
</h3>
<p>We've seen <a href="https://clojuredocs.org/clojure.core/deref"><code>deref</code></a> a couple of times now. And we're going to see it a bunch more times. So it's worth spending a minute or two more on it.</p>
<p>What does <code>deref</code> mean?</p>
<p>It stands for <strong>dereference</strong>. You see, both Promises and Futures are types of references. They're not the values themselves. They're pointers to the values. They're like boxes where the answer will be when the calculation is finished. You call <code>deref</code> on it and it gets whatever is in the box. We can also abbreviate <code>deref</code> by prepending a <code>@</code> to the reference, like this:</p>
<pre>
<code>@reference ;; same as (deref reference)</code>
</pre>
<p>Promises and Futures are references that might not be complete yet. In Clojure speak, when a promise has been delivered, we say it is <em>realized</em>. There's even a function called <code>realized?</code> to check if it has an answer. You can check if there's something in the box before you block getting the answer.</p>
<p>Whenever you block, it's also important sometimes to limit the amount of time you wait. Like in real life, you may want to say "if I don't hear from you by noon, I'm going to lunch by myself". You can use a variant of <code>deref</code> that has a timeout in it.</p>
<pre>
<code><span>;; wait four seconds (4,000 ms)</span>
<span>;; if we don't have an answer by then, deref will return :cheeseburger</span>
<span>(</span><span>deref</span> the-answer <span>4000</span> <span>:cheeseburger</span><span>)</span></code>
</pre>
<p>Only blocking variants of references can use the timeout version. For instance, Atoms, which we'll see real soon (I promise) always have a value. So there's no need to wait, ever.</p>
<p>Ok, let me deliver on that promise. &#128521;</p>
<p>I've made a video lesson about this. Check out <a href="https://purelyfunctional.tv/lesson/deref/"><em>Deref</em></a>.</p>
<h3>Atom</h3>
<p>Atoms are maybe the most popular concurrency primitive in Clojure right now. Don't quote me on that, because I don't have any data. But I find that I use them quite a lot more than other things.</p>
<p>Why do I use them so much?</p>
<p>Because they capture the essence of sharing a single piece of information.</p>
<p>Let's say I'm working with nine other people. We are collecting donations from people and sharing out how much we've collected. To work together, we keep a running sum on a chalkboard. John's job is to tweet out every five minutes how much money we have. He glances over at the chalkboard, sees the number "42", and tweets out this:</p>
<p>"We have collected a total of $42!!"</p>
<p>Except that's wrong. Jane was in the middle of writing out the number 42,332. She had to glance down to remember the last three digits, just when John was checking the board.</p>
<p>It sounds dumb when people do it, but this is how computers work. They're dumb. If you say "check the board and tweet out what you see", that's what you get.</p>
<p>But Atoms are a solution.</p>
<p>What's the real problem? The deep root of the problem is that you can write partial answers on the board. If you could write the whole number at once, this never would have happened. Atoms ensure that you can only write consistent values.</p>
<p>An Atom holds a current value. That value has to be valid. You then send it a pure function which calculates the next value. The Atom guarantees that anybody derefing the Atom always gets either the old value or the new value, and never anything in between.</p>
<p>Here's another problem. It's one you're probably familiar with:</p>
<p>Let's say Jane and Jim both collect $10 at about the same time. They both run back to the chalkboard and see this value:</p>
<p>$70,400</p>
<p>Great! They both pull out their calculators, type in the number, then add 10. Jane finishes first, erases the chalkboard, and write $70,410. She runs back out to get more donations. Then Jim erases the board. And writes $70,410. The same number!</p>
<p>They both were following a correct algorithm. Get the number, add your additions, then write it on the board. But when two people are involved, they need to coordinate a little better. Again, this is dumb when people are doing it, but it's what happens when threads share memory without coordination.</p>
<p>Let's see an example of using Atoms to prevent this problem.</p>
<pre>
<code><span>;; create an Atom with initially no money</span>
<span>(</span><span>def</span> donation-count <span>(</span>atom <span>0</span><span>)</span><span>)</span>

<span>;; start 9 people collecting money (9 threads)</span>
<span>(</span><span>dotimes</span> <span>[</span>_ <span>9</span><span>]</span>
  <span>(</span><span>doto</span> <span>(</span>Thread. <span>(</span><span>fn</span> <span>[</span><span>]</span>
                   <span>;; wait three seconds</span>
                   <span>(</span>Thread/sleep <span>3000</span><span>)</span>
                   <span>;; go collect $1</span>
                   <span>(</span>swap! donation-count <span>inc</span><span>)</span>
                   <span>;; do it again</span>
                   <span>(</span><span>recur</span><span>)</span><span>)</span><span>)</span>
    .start<span>)</span><span>)</span>

<span>;; start one person tweeting</span>
<span>(</span><span>doto</span> <span>(</span>Thread. <span>(</span><span>fn</span> <span>[</span><span>]</span>
                 <span>;; wait 100 seconds</span>
                 <span>(</span>Thread/sleep <span>100000</span><span>)</span>
                 <span>(</span>tweet <span>(</span><span>str</span> <span>"We collected $"</span> @donation-count <span>" total!"</span><span>)</span><span>)</span><span>)</span><span>)</span>
  .start<span>)</span></code>
</pre>
<p>There's our friend <code>deref</code> again. An Atom is a reference to a value, and derefing it gives you the value. But what's <a href="https://clojuredocs.org/clojure.core/swap!"><code>swap!</code></a>?</p>
<p><code>swap!</code> is the function for modifying the current value of the Atom. Let's look at the arguments:</p>
<pre>
<code><span>(</span>swap!
  donation-count <span>;; the Atom</span>
  <span>inc</span>            <span>;; the function</span>
  <span>)</span></code>
</pre>
<p><code>swap!</code> takes the current value of the Atom, calls the function on it (in this case, <code>inc</code>), and sets the new value. However, just before setting the new value, it checks to make sure the old value is still in there. If it's different, it starts over. It calls the function again with the new value it found. It keeps doing this until it finally writes the value. Because it can check the value and write it in one go, it's an atomic operation, hence the name.</p>
<p>That means the function can get called multiple times. That means it needs to be a <strong>pure function</strong>. Another thing is that you can't control the order of the function calls. If multiple threads are swapping to an Atom at the same time, order is out of the window. So make sure your functions are independent of order, like we talked about before.</p>
<p>For instance, incrementing (<code>inc</code>) is independent of order. After all the <code>inc</code>s are in, you'll have the same answer, regardless of order.</p>
<p>So, with all of those requirements, what benefits do you get from an Atom?</p>
<p>Let me explain.</p>
<p>Atoms give you a very important guarantee: you can look at the current value of the Atom (with <code>deref</code>) at any time and know that, at that time, the value was current. It was valid. It might not be current forever, but it was a valid answer. For instance, in our donations example, the value in the Atom is always the sum of some subset of donations. Any time you check the Atom, it was a valid count of money.</p>
<p>Let's look a bit more at the <code>swap!</code> function. We were calling <code>inc</code>, which adds one. What if you want to add more than just $1 each time?</p>
<pre>
<code><span>(</span>swap! donation-count <span>(</span><span>fn</span> <span>[</span>x<span>]</span> <span>(</span><span>+</span> x <span>10</span><span>)</span><span>)</span><span>)</span></code>
</pre>
<p>Okay, see, we can add 10, or whatever number you want. But this is such a common case that there's a shortcut. See, you're calling <code>+</code> with the current value of the Atom (here called <code>x</code>) and a second argument (<code>10</code>). Here's the same thing but using the shortcut:</p>
<pre>
<code><span>(</span>swap! donation-count <span>+</span> <span>10</span><span>)</span></code>
</pre>
<p>This shortcut trips people up. The arguments are in a funny order, but we can take it step by step:</p>
<pre>
<code><span>(</span>swap!
  donation-count <span>;; the Atom</span>
  <span>+</span>              <span>;; the function</span>
  <span>;; ;; current value of Atom goes here!</span>
  <span>10</span>             <span>;; the second argument</span>
  <span>;; ;; the third argument</span>
  <span>;; ;; the fourth argument, etc</span>
  <span>)</span></code>
</pre>
<p>I've used semicolons up there as placeholders for arguments. But we can see what's going on there: the current value of the Atom gets put as the first argument.</p>
<p>Okay, I'm going to wipe the sweat off my brow. Because next up we're looking at the biggest, most powerful reference type in Clojure.</p>
<p>I've made a video lesson about this. Check out <a href="https://purelyfunctional.tv/lesson/keeping-state-consistent-with-atoms/"><em>Keeping state consistent with Atoms</em></a>.</p>
<h3>Ref</h3>
<p>Listen, back in 2008/2009, everybody was talking about Refs. Why? Because they were Software Transactional Memory (STM), which was so hot in Clojure and one of the things that made it special. After a few years, it turns out that Atoms, which are much simpler, are good enough for most purposes. But the STM is still in there and people do use it.</p>
<p>When should you use it?</p>
<p>If we look at an Atom, it only holds one value. That value can be complex (a nested <a href="https://purelyfunctional.tv/guide/clojure-collections/">collection</a>, for instance), but it's just one value. If you have two Atoms, there's no way to make sure they stay in relationship. For example, you can't make one Atom always be twice the value of another. Why? Because some thread could read the Atoms after you change one but before you change the other.</p>
<p>Refs solve this problem. You can read from and modify multiple Refs inside of a transaction. Observers on the outside of the transaction cannot see the intermediate values. Nice!</p>
<p>Basically, Refs let you read and write to two chalkboards without any time passing between the first chalkboard and the second.</p>
<p>Let's look at some code:</p>
<pre>
<code><span>(</span><span>def</span> total-donations <span>(</span><span>ref</span> <span>0</span><span>)</span><span>)</span>

<span>(</span><span>def</span> count-donations <span>(</span><span>ref</span> <span>0</span><span>)</span><span>)</span>

<span>;; start 9 people collecting money</span>
<span>(</span><span>dotimes</span> <span>[</span>_ <span>9</span><span>]</span>
  <span>(</span><span>doto</span> <span>(</span>Thread. <span>(</span><span>fn</span> <span>[</span><span>]</span>
                   <span>;; go collect $10</span>
                   <span>;; ...</span>
                   <span>(</span><span>dosync</span>
                     <span>;; record $10</span>
                     <span>(</span><span>alter</span> total-donations <span>+</span> <span>10</span><span>)</span>
                     <span>;; record one donation</span>
                     <span>(</span><span>alter</span> count-donations <span>inc</span><span>)</span><span>)</span>
                   <span>;; do it again</span>
                   <span>(</span><span>recur</span><span>)</span><span>)</span><span>)</span>
    .start<span>)</span><span>)</span>

<span>;; start one person tweeting the total</span>
<span>(</span><span>doto</span> <span>(</span>Thread. <span>(</span><span>fn</span> <span>[</span><span>]</span>
                 <span>;; wait 100 seconds</span>
                 <span>(</span>Thread/sleep <span>100000</span><span>)</span>
                 <span>(</span>tweet <span>(</span><span>str</span> <span>"We collected $"</span> @total-donations <span>" total!"</span><span>)</span><span>)</span><span>)</span><span>)</span>
  .start<span>)</span>

<span>;; start one person tweeting the average</span>
<span>(</span><span>doto</span> <span>(</span>Thread. <span>(</span><span>fn</span> <span>[</span><span>]</span>
                 <span>;; wait 100 seconds</span>
                 <span>(</span>Thread/sleep <span>100000</span><span>)</span>
                 <span>(</span><span>when</span> <span>(</span><span>pos?</span> @count-donations<span>)</span>
                   <span>(</span>tweet <span>(</span><span>str</span> <span>"Average donation: $"</span> <span>(</span><span>double</span> <span>(</span><span>dosync</span> <span>(</span><span>/</span> @total-donations @count-donations<span>)</span><span>)</span><span>)</span><span>)</span><span>)</span><span>)</span><span>)</span><span>)</span>
  .start<span>)</span></code>
</pre>
<p><a href="https://clojuredocs.org/clojure.core/dosync"><code>dosync</code></a> means "do synchronized". It means that everything in there is within a <em>transaction</em>. Transactions give you some guarantees: if any funny business happens in that transaction (an exception is thrown, for example), the transaction will be aborted like it never happened. Inside of that transaction, you have a consistent view of all of the Refs you dereference. Any changes you make are visible to you from the inside, but not to others outside of the transaction, until the transaction completes. It's like you're in a little time bubble.</p>
<p>As a cost for these guarantees, you have to obey a few rules. First, no side effects in a transaction. The transaction can be run multiple times. Also notice that like Atoms, you can't guarantee the order.</p>
<p>Also notice that instead of using <code>swap!</code>, you use <a href="https://clojuredocs.org/clojure.core/alter"><code>alter</code></a>. But it's very similar. You can even do the extra arguments thing like we do above.</p>
<p>I've made a video lesson about this. Check out <a href="https://purelyfunctional.tv/lesson/ref/"><em>Ref</em></a>.</p>
<h3>Var</h3>
<p>We need to talk about Vars a little. But before I do, let me say this: you almost never use Vars directly as a concurrency primitive. So I'm not going to go deep into them. But they're important. We have to talk about them.</p>
<p>Whenever you define a variable using <code>def</code> or <code>defn</code>, you create a Var. They're references like Refs and Atoms. That means they're mutable. You can change the value of a Var. And we use that all the time while we're doing <strong>interactive development</strong>. We define a function, we realize it's not quite right, so we redefine it. That redefinition changes what's called the <strong>root value</strong> of the Var.</p>
<p>In addition to the root value, Vars can have a different value per-thread. This lets different threads use different <strong>dynamic scopes</strong> with the same Var.</p>
<p>Vars are one of those things that recede into the background. You rarely use them explicitly. They are there only to support interactive development and dynamic scope.</p>
<p>I've made a video lesson about this. Check out <a href="https://purelyfunctional.tv/lesson/vars/"><em>Vars</em></a>.</p>
<h3>Agent</h3>
<p>For some reason, whenever you mention Agents, people think of <strong>Actors</strong>, which are found in Erlang. They're definitely not the same. An Actor receives messages, does some work based on which message they receive, then listens for more messages. Agents, on the other hand, hold state like an Atom or a Ref.</p>
<p>It's best to compare Agents to Atoms. Like Atoms, Agents are <em>uncoordinated</em>. You can't modify two Agents with any kind of guarantees. The difference is all about which thread does the work. When you <code>swap!</code> an Atom, the processing happens in the current thread. The thread keeps retrying the computation until it successfully saves to the Atom (or throws an exception). But everything you do to an Atom happens in the current thread.</p>
<p>Calling <a href="https://clojuredocs.org/clojure.core/send"><code>send</code></a> on an Agent, on the other hand, runs the computation on another thread. The call to <code>send</code> returns immediately after adding a job to a work queue that will be processed by a thread pool. So it's like an Atom, but stuff happens on another thread.</p>
<p>Let's look at the interface to Agent. <code>send</code> is the Agent equivalent of <code>swap!</code> and <code>alter</code>. Like I said, <code>send</code> will process the computation in a thread pool. You use thread pools for quick tasks, like adding or string concatenation. If your task will take a long time--a long computation or I/O--you could quickly overwhelm all of the threads in the pool and keep them too busy to take on more work. If your function does do lots of work, Clojure gives you a function called <a href="https://clojuredocs.org/clojure.core/send-off"><code>send-off</code></a>, which runs each task in its own thread. Use that if you're doing I/O or a lot of computation.</p>
<p>Let's write a simple summing function that stores the answer in an Agent.</p>
<pre>
<code><span>(</span><span>def</span> sum <span>(</span><span>agent</span> <span>0</span><span>)</span><span>)</span> <span>;; create an agent initialized to 0</span>

<span>(</span><span>def</span> numbers <span>[</span><span>0</span> <span>9</span> <span>3</span> <span>4</span> <span>5</span> <span>5</span> <span>4</span> <span>44</span> <span>4</span> <span>2</span> <span>5</span> <span>6</span> <span>7</span> <span>775</span> <span>..</span>.<span>]</span><span>)</span>

<span>(</span><span>doseq</span> <span>[</span>x numbers<span>]</span>
  <span>(</span><span>send</span> sum <span>+</span> x<span>)</span><span>)</span> <span>;; add x to the current value</span>

<span>(</span><span>await</span> sum<span>)</span> <span>;; wait until all sent actions are done</span>

<span>(</span><span>println</span> @sum<span>)</span> <span>;; should have the answer</span></code>
</pre>
<p>It might be easy to think that this happens in parallel. Even though it's happening on multiple threads, it's not parallel. Each Agent has its own queue of tasks, and they are done in the order they are received, one at a time. So it's not parallel. If you want parallelism with Agents, you have to have many Agents.</p>
<p>How can we make this sum parallel? Easy. Just make multiple Agents in a loop.</p>
<pre>
<code><span>;; make 10 agents initialized to zero</span>
<span>(</span><span>def</span> sums <span>(</span><span>map</span> <span>agent</span> <span>(</span><span>repeat</span> <span>10</span> <span>0</span><span>)</span><span>)</span><span>)</span>

<span>(</span><span>def</span> numbers <span>(</span><span>range</span> <span>1000000</span><span>)</span><span>)</span> <span>;; one million numbers</span>

<span>;; loop through all numbers and round-robin the agents</span>
<span>(</span><span>doseq</span> <span>[</span><span>[</span>x <span>agent</span><span>]</span> <span>(</span><span>map</span> <span>vector</span> numbers <span>(</span><span>cycle</span> sums<span>)</span><span>)</span><span>]</span>
  <span>(</span><span>send</span> <span>agent</span> <span>+</span> x<span>)</span><span>)</span>

<span>;; wait at most 10 seconds</span>
<span>(</span><span>apply</span> <span>await-for</span> <span>10000</span> sums<span>)</span>

<span>;; sum up the answers in all ten agents</span>
<span>(</span><span>println</span> <span>(</span><span>apply</span> <span>+</span> <span>(</span><span>map</span> <span>deref</span> sums<span>)</span><span>)</span><span>)</span></code>
</pre>
<p>Of course, summing is just an example. It's probably not worth queuing up a task that just adds (addition is faster than queuing). But this shows how to do any kind of work in parallel.</p>
<p>In this example, all agents get the same number of tasks. What happens if some tasks take longer than others? That means that some Agents will be idle while others are still working. How can you prevent that?</p>
<p>The answer, predictably, is to add another level of indirection.</p>
<p>Instead of round-robin, we should add our numbers to a queue that the agents pull from when they're ready for more work.</p>
<pre>
<code><span>;; make 10 agents initialized to zero</span>
<span>(</span><span>def</span> sums <span>(</span><span>map</span> <span>agent</span> <span>(</span><span>repeat</span> <span>10</span> <span>0</span><span>)</span><span>)</span><span>)</span>

<span>(</span><span>def</span> numbers <span>(</span><span>agent</span> <span>(</span><span>range</span> <span>1000000</span><span>)</span><span>)</span><span>)</span> <span>;; one million numbers in an agent</span>

<span>(</span><span>defn</span> dequeue-and-<span>add</span> <span>[</span>sum-agent<span>]</span>
  <span>(</span>letfn <span>[</span><span>(</span><span>add</span> <span>[</span>current-sum x<span>]</span>
            <span>;; do the addition</span>
            <span>(</span><span>let</span> <span>[</span>new-sum <span>(</span><span>+</span> current-sum x<span>)</span><span>]</span>
              <span>;; when we're done, schedule the next dequeue</span>
              <span>(</span><span>send</span> numbers dequeue<span>)</span>
              <span>;; return the new value of the Agent</span>
              new-sum<span>)</span><span>)</span>
          <span>(</span>dequeue <span>[</span>xs<span>]</span>
              <span>;; check if there's more to do</span>
              <span>(</span><span>when</span> <span>(</span><span>not</span> <span>(</span>empty? xs<span>)</span><span>)</span>
                <span>;; send the first number to the Agent</span>
                <span>(</span><span>send</span> sum-agent <span>add</span> <span>(</span><span>first</span> xs<span>)</span><span>)</span><span>)</span>
                <span>;; return the other numbers for other Agents</span>
                <span>(</span><span>rest</span> xs<span>)</span><span>)</span><span>]</span>
    <span>(</span><span>send</span> numbers dequeue<span>)</span><span>)</span><span>)</span>

<span>;; start all 10 Agents working</span>
<span>(</span><span>doseq</span> <span>[</span>sum-agent sums<span>]</span>
  <span>(</span>dequeue-and-<span>add</span> sum-agent<span>)</span><span>)</span>

<span>;; wait for all the numbers to be cleared from the queue</span>
<span>(</span><span>loop</span> <span>[</span><span>]</span>
  <span>(</span><span>when</span> <span>(</span><span>seq</span> @numbers<span>)</span>
    <span>(</span>Thread/sleep <span>1000</span><span>)</span>
    <span>(</span><span>recur</span><span>)</span><span>)</span><span>)</span>

<span>;; sum up the answers in all ten agents</span>
<span>(</span><span>println</span> <span>(</span><span>apply</span> <span>+</span> <span>(</span><span>map</span> <span>deref</span> sums<span>)</span><span>)</span><span>)</span></code>
</pre>
<p>We're using a <strong>hand off</strong> pattern. In <code>dequeue-and-add</code> above, <code>add</code> triggers <code>dequeue</code> and <code>dequeue</code> triggers <code>add</code>. <code>dequeue</code> takes a number from the <code>numbers</code> Agent and <code>add</code> adds it to the summing Agent.</p>
<p>We can't use <a href="https://clojuredocs.org/clojure.core/await"><code>await</code></a> on this one because the Agents use a hand off pattern. There will be times when there are no tasks queued on an Agent because it is waiting for another task to finish on another Agent. We have to be clever and wait for the queue to process. You should probably add a timeout to that, in case the queue never does empty.</p>
<p>What about errors?</p>
<p>Well, that's really interesting. If your task throws an exception, it gets stored inside the Agent (not as the Agent's state). Once there's an exception, no more tasks will be processed. You can check if an Agent has failed by calling <a href="https://clojuredocs.org/clojure.core/agent-error"><code>agent-error</code></a> on it. It will return the exception (or <code>nil</code> if there isn't one). Then you can clear it with <a href="https://clojuredocs.org/clojure.core/restart-agent"><code>restart-agent</code></a>. You can have the Agent automatically handle its own errors with <a href="https://clojuredocs.org/clojure.core/set-error-handler!"><code>set-error-handler!</code></a>. You pass it a function of two arguments (the agent and the exception) and you can handle the error.</p>
<p>Agents are really flexible, but as you can see, the code can get kind of complicated. For doing stuff in parallel, I'd probably use other options, like reducers (for pure functions) and <code>ExecutorService</code> (for blocking tasks).</p>
<p>Agents are nice for things where you have a value that needs to be accumulated to over time and you want to do a calculation on it in another thread. They can also do side effects in the function you send them, because they're queued up instead of competing.</p>
<p>I've made a video lesson about this. Check out <a href="https://purelyfunctional.tv/lesson/agents/"><em>Agents</em></a>.</p>
<h3>Lock</h3>
<p>Locks are the traditional, old-school way of coordinating access to resources. Like a lock on your bathroom, software locks make sure only one person is using that resource (the bathroom) at the same time. We call that <strong>mutual exclusion</strong>: if I'm in the bathroom, you can't be in here. I exclude you and you exclude me. While a thread has a lock, it can act like it's the only thread that has that resource. It's a very low-level way of coordinating, but sometimes that's exactly what you want.</p>
<p>All JVM objects have a lock built in. You usually don't notice, but it's how Java implements the <code>synchronized</code> keyword. So all you need to create a lock is to create an <code>Object</code>.</p>
<p>Let's solve an actual problem using locks. If you have many threads all printing to the console at the same time, very often you'll see that the lines are mixed up. Two threads that print at exactly the same time will send their characters at the same time, and the line is just messed up.</p>
<pre>
<code><span>(</span><span>defn</span> log <span>[</span>&amp; args<span>]</span>
  <span>(</span><span>apply</span> <span>println</span> args<span>)</span><span>)</span>

<span>;; Thread 1</span>
<span>(</span>log <span>"INFO 2017-4-29: Starting database connection."</span><span>)</span>

<span>;; Thread 2</span>
<span>(</span>log <span>"WARNING 2017-4-29: Cannot find configuration file, using defaults."</span><span>)</span></code>
</pre>
<p>Instead of the following, which is what you want:</p>
<pre>
<code>INFO 2017-4-29: Starting database connection.
WARNING 2017-4-29: Cannot find configuration file, using defaults.</code>
</pre>
<p>You get something like this:</p>
<pre>
<code>INFO 20WARN17-4-29: StartingING 2017-4-29: Cannot find configuration file, database connection.</code>
</pre>
<p>The characters from the two <code>println</code>s got mixed up. They were being sent to standard out at the same time from two different Threads.</p>
<p>What's the solution?</p>
<p>Lock an object so that only one Thread can be in some code at the same time.</p>
<pre>
<code><span>;; construct an Object just for its lock</span>
<span>(</span><span>def</span> log-lock <span>(</span>Object.<span>)</span><span>)</span>

<span>(</span><span>defn</span> log <span>[</span>&amp; args<span>]</span>
  <span>(</span><span>locking</span> log-lock
    <span>(</span><span>apply</span> <span>println</span> args<span>)</span><span>)</span><span>)</span></code>
</pre>
<p>Now you can call <code>log</code> in many different Threads. The lines will always come out okay.</p>
<p>Like I said, <code>locking</code> is a low-level tool. I don't know if I've ever used it in production. But if you have some resource shared by multiple Threads, this is a simple way to let them share it safely.</p>
<p>I've made a video lesson about this. Check out <a href="https://purelyfunctional.tv/lesson/locking/"><em>Locking</em></a>.</p>
<h3>core.async</h3>
<p>If you've written Clojure, you've probably heard of <a href="https://github.com/clojure/core.async">core.async</a>. core.async is a great library for doing parallel processing. The reason is simple: it's a very lightweight way to break down tasks and communicate between the tasks.</p>
<p>You create a task by using the <a href="https://clojuredocs.org/clojure.core.async/go"><code>go</code></a> macro. All of the code inside of a <code>go</code> block will be executed in a <strong>Process</strong>. Processes are lighter weight than Threads and so are better for breaking stuff up even more than you normally would with Threads.</p>
<p>Processes can communicate using lightweight queues called <strong>Channels</strong>. The <a href="https://purelyfunctional.tv/guide/clojure-collections/#patterns">patterns</a> are endless, but just as an example, you could create one Process that puts values onto a Channel and another Process that consumes them. Channels ensure that Processes wait for each other and that each value on the Channel is delivered only once.</p>
<p>Let's look at some code:</p>
<pre>
<code><span>(</span>require <span>[</span>clojure.core.async <span>:as</span> async<span>]</span><span>)</span>

<span>;; create a channel with a buffer of up to 100 values</span>
<span>(</span><span>def</span> number-chan <span>(</span>async/chan <span>100</span><span>)</span><span>)</span>
<span>;; Atom where we keep the sum</span>
<span>(</span><span>def</span> sum <span>(</span>atom <span>0</span><span>)</span><span>)</span>

<span>;; start 100 go processes taking numbers from number-chan</span>
<span>;; and adding them to the sum Atom</span>
<span>(</span><span>dotimes</span> <span>[</span>_ <span>100</span><span>]</span>
  <span>(</span>go
    <span>(</span><span>loop</span> <span>[</span><span>]</span>
      <span>(</span><span>let</span> <span>[</span>number <span>(</span>async/<span>&lt;</span>! number-chan<span>)</span><span>]</span>
        <span>(</span>swap! sum <span>+</span> number<span>)</span><span>)</span>
      <span>(</span><span>recur</span><span>)</span><span>)</span><span>)</span><span>)</span>

<span>;; create a go process that adds the numbers 0-1 million</span>
<span>;; to the channel</span>
<span>(</span>go
  <span>(</span><span>doseq</span> <span>[</span>x <span>(</span><span>range</span> <span>1000000</span><span>)</span><span>]</span>
    <span>(</span>async/<span>&gt;</span>! number-chan x<span>)</span><span>)</span><span>)</span></code>
</pre>
<p>You take from a Channel with the function <a href="https://clojuredocs.org/clojure.core.async/%3C!"><code>&lt;!</code></a> and put to a Channel with <a href="https://clojuredocs.org/clojure.core.async/%3E!"><code>&gt;!</code></a>. If you try to take and there is no value, your Process will "park" and wait for a value. If you put and there is no room for new values, your Process will "park" and wait for room on the Channel. Because things wait, it means you can coordinate. You can say "Let's each go do some work and meet back here when we're done."</p>
<p>core.async uses <a href="https://en.wikipedia.org/wiki/Communicating_sequential_processes">Communicating Sequential Processes</a> (CSP), which is an algebra for building concurrent systems. It's the same abstraction that the <a href="https://golang.org/">Go programming language</a> uses.</p>
<p>This is just a very short introduction to a very big topic. You can learn more about core.async from these resources:</p>
<ul>
<li>
<a href="https://www.infoq.com/presentations/clojure-core-async">Clojure core.async</a>: a presentation by Rich Hickey at Strange Loop.</li>
<li>
<a href="http://pftv.wpengine.com/courses/clojure-core-async/">Clojure core.async</a>: my course on the topic.</li>
<li>
<a href="http://pftv.wpengine.com/courses/core-async-patterns/">core.async Patterns</a>: an advanced course where you learn interesting patterns for using core.async.</li>
<li>
<p><a href="http://www.braveclojure.com/core-async/">Mastering Concurrent Processes with core.async</a>: The excellent chapter from <em>Clojure for the Brave and True</em>.</p>
<p>I've made a video lesson about this. Check out <a href="https://purelyfunctional.tv/lesson/core-async/"><em>Core Async</em></a>.</p>
</li>
</ul>
<h3>Threads</h3>
<p>You can create Threads in Java very easily by constructing the <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Thread.html"><code>java.lang.Thread</code></a> class. Java has the concept of a <code>Runnable</code>, which is an interface with a method called <code>run()</code> that returns nothing. If you want to run something in another thread, you pass a <code>Runnable</code> to the constructor of a <code>Thread</code>, then start it. Luckily, Clojure has thought of this. You can use a function of zero arguments directly as a <code>Runnable</code>.</p>
<pre>
<code><span>;; create a Thread</span>
<span>(</span><span>def</span> thread <span>(</span>Thread. <span>(</span><span>fn</span> <span>[</span><span>]</span> <span>;; 0 arguments</span>
                       <span>;; this will run in a new Thread</span>
                       <span>(</span><span>println</span> <span>1</span> <span>2</span> <span>3</span><span>)</span><span>)</span><span>)</span><span>)</span>

<span>;; the thread won't run until you start it</span>
<span>(</span>.start thread<span>)</span></code>
</pre>
<p>Threads are easy to use in Java, but there are some things to keep in mind. First of all, Threads cannot be stopped from the outside. The only way to stop a Thread is for the function you pass it to reach the end of execution. That means if you create an infinite loop, that Thread will run forever. Just be aware!</p>
<p>So how do you stop a JVM Thread? You have to code it to watch for a signal. Maybe you set up a Promise that it checks. When the Promise is delivered, the Thread will stop executing.</p>
<p>Which brings us to the second point, there is no built-in way to communicate with a Thread. The Threads share memory, so they have access to all of the objects in scope. So you can use something like a Promise or a core.async channel.</p>
<p>Finally, there's no way to communicate out of the Thread. The return value of your function is discarded. If you do need to get a value out, you can put the value in a Promise, or otherwise store it in a Ref or Atom. For heavy-duty communication between Threads, consider core.async.</p>
<p>Wait. There's one more thing:</p>
<p>The JVM won't exit until all Threads are finished. When you're running Threads and you want to shut down the JVM, you have to call <code>System.exit()</code>. In Clojure, that looks like this:</p>
<pre>
<code><span>(</span>System/exit <span>0</span><span>)</span> <span>;; 0 means success, it's the Unix exit code</span></code>
</pre>
<ul>
<li>
<a href="https://docs.oracle.com/javase/tutorial/essential/concurrency/runthread.html">Defining and Starting a Thread</a>: A tutorial from Oracle.</li>
<li>
<p><a href="http://pftv.wpengine.com/jvm-clojure/java-threads/">Java Threads</a>: A lesson in my JVM course about starting and stopping threads.</p>
<p>I've made a video lesson about this. Check out <a href="https://purelyfunctional.tv/lesson/threads/"><em>Threads</em></a>.</p>
</li>
</ul>
<h3>Excutor Service</h3>
<p>The Java standard library contains something called an <code>ExecutorService</code>. If you have a bunch of similar tasks that need to be run in parallel, <code>ExecutorService</code> is your friend.</p>
<p>What does it do?</p>
<p>You set up a thread pool and a queue feeding that thread pool. Those make up your <code>ExecutorService</code>. You hand the <code>ExecutorService</code> "tasks". The tasks get queued up and pulled off by the threads in the pool, executing them in parallel.</p>
<p>Tasks are just instances of <code>Callable</code>. Since in Clojure functions are instances of <code>Callable</code>, it's super easy to just pass it functions of no arguments.</p>
<p>Let's see some code:</p>
<pre>
<code><span>(</span><span>import</span> <span>'java</span>.util.concurrent.ExecutorService<span>)</span>
<span>(</span><span>import</span> <span>'java</span>.util.concurrent.Executors<span>)</span>

<span>;; create a thread pool with 4 threads</span>
<span>(</span><span>def</span> service <span>(</span>Executors/newFixedThreadPool <span>4</span><span>)</span><span>)</span>

<span>;; submit a task and save the Future</span>
<span>(</span><span>def</span> <span>f</span> <span>(</span>.submit ^ExecutorService service    <span>;; we need to hint to tell it</span>
                ^Callable <span>(</span><span>fn</span> <span>[</span><span>]</span>            <span>;; to use the Callable version</span>
                            <span>;; do some work</span>
                            <span>..</span>.<span>)</span><span>)</span><span>)</span>

<span>;; block on the Future</span>
<span>(</span><span>println</span> @<span>f</span><span>)</span></code>
</pre>
<p>Clojure Futures are also run in a similar way. If you need custom control of the thread pool, you can do it yourself with an <code>ExecutorService</code>.</p>
<p>I've made a video lesson about this. Check out <a href="https://purelyfunctional.tv/lesson/executorservice/"><em>ExecutorService</em></a>.</p>
<h3>Manifold</h3>
<p><a href="https://github.com/ztellman/manifold">Manifold</a> is an interesting library from Zach Tellman. It is used as part of the pervasive asynchrony found in <a href="http://aleph.io/">Aleph</a>, the high-throughput, asynchronous networking library.</p>
<p>The interesting thing about Manifold is that it provides another level of indirection which captures the essence of Futures, Promises, core.async Channels, and even RxJava and ReactiveStreams.</p>
<p>If you're going to be doing a lot of asynchronous communication between your threads, Manifold could be for you.</p>
<h3><code>pmap</code></h3>
<p>I had to include <a href="https://clojuredocs.org/clojure.core/pmap"><code>clojure.core/pmap</code></a> here, just for completeness. It's a parallel implementation of <code>clojure.core/map</code>. However, you should note that it's very na&iuml;ve. It is still lazy, like regular <code>map</code>, so it won't begin executing until a value is needed. It keeps just ahead. However, the function you pass it does get run in other Threads. If it is computationally intensive, you may want to try <code>pmap</code>. Add one more letter and it makes your code parallel.</p>
<h3>Reducers</h3>
<p><a href="https://clojure.org/reference/reducers">Clojure reducers</a> is a great library for executing things in parallel. Under the hood, it uses Fork/Join. It also is custom-tuned to work with all of Clojure's built-in data structures.</p>
<p>To execute things in parallel with reducers, use <code>clojure.core.reducers/fold</code>. It is like <code>reduce</code> except it has two functions: one for reducing and one for combining. The collection you pass it will be broken into chunks. Each chunk will be reduced with the reducing function. The results are combined using the combining function.</p>
<pre>
<code><span>(</span>require '<span>[</span>clojure.core.reducers <span>:as</span> r<span>]</span><span>)</span>

<span>(</span><span>def</span> numbers <span>(</span>vec <span>(</span><span>range</span> <span>1000000</span><span>)</span><span>)</span><span>)</span>

<span>;; sum numbers in parallel</span>
<span>;; when combining function and reducing function are the same,</span>
<span>;; you can use this arity</span>
<span>(</span>r/fold <span>+</span> numbers<span>)</span>

<span>;; sum only even numbers</span>
<span>;; r/filter does not create intermediate lists</span>
<span>(</span>r/fold <span>+</span> <span>(</span>r/<span>filter</span> even? numbers<span>)</span><span>)</span></code>
</pre>
<p>The reducers library comes with Clojure. The main benefit is that <code>fold</code> can run using Fork/Join, and so efficiently use many cores. In addition, the standard sequence operations like <code>map</code>, <code>filter</code>, <code>drop</code>, <code>take</code>, and <code>cat</code> (concatenate) do not create intermediate sequences as they do with the <code>clojure.core</code> versions. They are specifically built to work in parallel using Fork/Join.</p>
<p>I've made a video lesson about this. Check out <a href="https://purelyfunctional.tv/lesson/reducers/"><em>Reducers</em></a>.</p>
<h3>Fork/Join</h3>
<p>Another system built into the Java standard library is called <a href="https://docs.oracle.com/javase/tutorial/essential/concurrency/forkjoin.html">Fork/Join</a>. Fork/Join is a way to break up a job into tiny tasks that get distributed to all of your cores. Fork/Join then reassembles the pieces into an answer.</p>
<p>I've made a video lesson about this. Check out <a href="https://purelyfunctional.tv/lesson/forkjoin/"><em>ForkJoin</em></a>.</p>
<p>Let's write our own summation system again.</p>
<pre>
<code><span>(</span><span>import</span> <span>'java</span>.util.concurrent.RecursiveTask<span>)</span>
<span>(</span><span>import</span> <span>'java</span>.util.concurrent.ForkJoinPool<span>)</span>

<span>(</span><span>def</span> pool <span>(</span>ForkJoinPool.<span>)</span><span>)</span>

<span>(</span><span>defn</span> summation <span>[</span>numbers<span>]</span>
  <span>(</span><span>proxy</span> <span>[</span>RecursiveTask<span>]</span> <span>[</span><span>]</span>
    <span>(</span>compute <span>[</span><span>]</span>
      <span>(</span><span>if</span> <span>(</span><span>&lt;=</span> <span>(</span><span>count</span> numbers<span>)</span> <span>512</span><span>)</span>
        <span>;; if the vector is small enough,</span>
        <span>;; we just reduce over them</span>
        <span>(</span><span>reduce</span> <span>+</span> <span>0</span> numbers<span>)</span>
        <span>;; otherwise, we split the vector roughly in two</span>
        <span>;; and recursively run two more tasks</span>
        <span>(</span><span>let</span> <span>[</span>half <span>(</span><span>quot</span> <span>(</span><span>count</span> numbers<span>)</span> <span>2</span><span>)</span>
              <span>f1</span> <span>(</span>summation <span>(</span><span>subvec</span> numbers <span>0</span> half<span>)</span><span>)</span>
              <span>f2</span> <span>(</span>summation <span>(</span><span>subvec</span> numbers half<span>)</span><span>)</span><span>]</span>
          <span>;; do half the work in a new thread</span>
          <span>(</span>.fork <span>f2</span><span>)</span>
          <span>;; do the other half in this thread and combine</span>
          <span>(</span><span>+</span> <span>(</span>.compute <span>f1</span><span>)</span> <span>(</span>.<span>join</span> <span>f2</span><span>)</span><span>)</span><span>)</span><span>)</span><span>)</span><span>)</span><span>)</span>

<span>(</span><span>defn</span> sum <span>[</span>numbers<span>]</span>
  <span>(</span>.invoke pool <span>(</span>summation <span>(</span>vec numbers<span>)</span><span>)</span><span>)</span><span>)</span>

<span>(</span><span>def</span> answer <span>(</span>sum <span>(</span><span>range</span> <span>1000000</span><span>)</span><span>)</span><span>)</span></code>
</pre>
<p>If we sum a great big list of numbers, it will get split into two halves until the list is 512 items or less. When we split, we recursively call <code>.fork</code> on one half, which will queue up the task for another thread. Meanwhile, since we have the thread, we can continue computing the first half. <code>.join</code> will block on the second half that is executing in another thread.</p>
<p>I know this isn't the best way to sum numbers, since addition doesn't really care about order. But it is a good way to split up work done on an <strong>associative</strong> operation. You will need to tune the size of the work done in one thread to make it worth it to fork. In general, you probably will want to use Clojure reducers for this kind of work. See below.</p>
<h2>The secret to understanding concurrency and parallelism</h2>
<p>A common question is "What is the difference between concurrency and parallelism?". If it's confusing you, don't worry. They're almost the same and the distinction is mostly academic. However, you're probably already used to thinking about concurrency and parallelism. We actually encounter the difference all the time in the real world.</p>
<p><strong>Let's look at an example: bank tellers.</strong></p>
<p>Imagine you walk into a bank. You see a row of tellers and a line of people waiting. How many bank transactions can happen at the same time? Easy. It's the same as the number of tellers. The <strong>parallelism</strong> of the bank is how many things can happen at the exact same time. If there are four tellers, four things can happen at the same time.</p>
<p>Even though there are only four tellers, all of those clients waiting in line will be helped before the end of the day. They're basically competing over the scarce resource of the teller's time. But because of the <strong>concurrency</strong> system set up (the queue), they all know that their business will be handled eventually.</p>
<h3>Definitions</h3>
<p><strong><em>concurrency</em></strong>: <em>ability to share resources (such as CPUs, RAM, databases, etc.) safely and efficiently</em></p>
<p><strong><em>parallelism</em></strong>: <em>adding more resources to get the answer faster</em></p>
<p>Thanks to <a href="https://www.youtube.com/watch?v=NsDE7E8sIdQ&amp;t=514s">Brian Goetz</a> for these definitions.</p>
<p><strong>Want another example?</strong></p>
<p>How about a bathroom shared among roommates. How many people can use it at once? One. But somehow using a lock, the eight roommates can share it.</p>
<p>We can increase the parallelism by adding bathrooms. If you've got ten bathrooms but only eight roommates, we've actually got more parallelism than concurrency. The work of using the bathroom can't really be broken down any more to take advantage of those two extra rooms. You can't go to the bathroom faster by using two bathrooms.</p>
<p><strong>Let's go back down to one bathroom.</strong></p>
<p>With eight roommates sharing one bathroom, can you imagine maybe one roommate doesn't get their fair share of bathroom time? Maybe they are slower running to the bathroom when it's vacant. Those roommates could use a better system for sharing that bathroom than a simple lock. This is where <strong>concurrency primitives</strong> come in. Concurrency primitives are simple tools that you can use to build concurrent systems that have the properties you need to solve your problem.</p>
<p>Want some real-world examples of concurrency primitives?</p>
<ul>
<li>Locks</li>
<li>Queues</li>
<li>Schedules</li>
<li>Conveyor belts</li>
</ul>
<p>We'll see more in a bit, where we'll catalog the ones that you get with Clojure.</p>
<p>But first, we saw how to increase parallelism. But how do you increase concurrency? We already know you can't increase concurrency in the bathroom example. The "unit of work" can't be split up any more. One person can't use two bathrooms at the same time!</p>
<p>But you can do it in the bank example.</p>
<p>Let's say you have 100 bank tellers (100 parallelism) and 10 clients in line. Obviously, no one will have to wait--except for those 90 tellers who are not busy! Is there anything we can do to keep them busy to make the client's bank experience go faster?</p>
<p>Yes. Because if you ask those clients, each one has multiple transactions. One person has four checks to deposit. Another wants to do three separate transfers, etc. The bank has set up their work so that each transaction is entirely independent. You can hand each teller a check to deposit (with the deposit slip) and all of your work can be done at the same time. One way to increase concurrency is to break down the work to be done into independent chunks. You'd be surprised how much can be broken up without much work.</p>
<p>For instance:</p>
<p>Let's say you have to fetch 20 web pages. Each GET request is independent of the others, so you can do them all at the same time. No problem.</p>
<p>That's an easy example of independent work. We'll get deeper into what it means to be "independent" later, because there are different types of independence.</p>
<p>But that's for later.</p>
<p>Let's dive into the computer world and see the basics that Clojure and the JVM give us.</p>
<h2>Threads and the JVM</h2>
<p>In one computer, the parallelism is the number of cores. Each core can execute instructions independently and run at full speed. The JVM does not give you direct access to these cores. You have to go through the operating system, which gives you processes and threads. These are both concurrency mechanisms so that different programs can share the limited number of cores.</p>
<p>The JVM lets you create native operating system threads. All of the threads share memory.</p>
<p>So let's look at this more closely:</p>
<p>As far as the JVM is concerned, you can have unlimited threads. And they're all able to read and write to the same memory. Obviously, all of those threads will need some help working together.</p>
<p>So let's give them some help!</p>
<p>Let's get into the concurrency tools we've got in Clojure.</p>
<h2>Immutable data structures</h2>
<p>The first one is kind of not that obvious because it's everywhere in Clojure. But it's one of those important things that makes all of the other concurrency stuff possible.</p>
<p>What is it?</p>
<p>The immutable data structures. Clojure uses immutable data by default. Why is this important? Let's imagine a bank where you could reuse checks. Let's write all the information in pencil. That <strong>could</strong> work, with a lot of discipline. Or you could make checks one-time-use (you create one, use it, then throw it away).</p>
<p>Or what if we wrote the amount of money in your account in chalk on a chalkboard. Whenever you deposit a check, we just erase the old number and write the new number. Sounds great, right?</p>
<p>No! It sounds terrible. Change banks immediately!</p>
<p>Or how about a bathroom shared by eight people with no lock! You can do it. Just be disciplined. Knock every time. What could go wrong? Start looking for a new place to live (or run to the hardware store to buy a lock).</p>
<p>It sounds ridiculous, but that's basically what we do all the time when we use mutable data structures. You can still write concurrent systems, but your job is harder and your success depends on discipline instead of easy rules like "wait your turn in line".</p>
<p>So let's all just take a moment and appreciate the calming and concurrent nature of immutable data structures.</p>
<p>Now we can move on to the next tool that really helps us.</p>
<h2>Pure functions</h2>
<p>The next thing that really helps with concurrency is using pure functions. Clojure doesn't help much with this, except that it makes it easy to make pure functions. But you don't get pure functions by default. And Clojure can't tell you whether a function is pure.</p>
<p>What is a pure function?</p>
<p>A pure function gives you the same return value for the same inputs all the time. You can call it whenever you want as many times as you want and you'll always get the same return value.</p>
<p>Wait! What things can make a function return different things?</p>
<p>What if the function reads a global mutable variable?</p>
<p>What if the function asks for user input?</p>
<p>What if the function fetches something from the network?</p>
<p>What if the function does different things depending on the time?</p>
<p>Or what if the function generates a different random number each time?</p>
<p>Basically, they can't depend on anything that changes. The time changes. The network changes. The user changes.</p>
<p>There's one more thing about pure functions you have to know about: not only can they not depend on changing stuff, pure functions can't change anything themselves.</p>
<p>Let's look at some examples of changing stuff:</p>
<p>Printing to the screen.</p>
<p>Posting to a web server.</p>
<p>Writing to disk.</p>
<p>Changing a global variable.</p>
<p>Sending an email.</p>
<p>And the classic: launching a missile. &#128640;</p>
<p>Why are pure functions important?</p>
<p>Because they're easy to understand. They don't depend on the history of the program. They always act the same. They're like a rock solid friend--you can always trust them.</p>
<p>But there's another reason.</p>
<p>Many concurrency primitives will sometimes call your function multiple times. They need to. Why?</p>
<p>Let's look at an example.</p>
<p>Imagine you're working with a stock market agent. You've got some money and want to invest. The agent comes to you, tells you how much money you have and the current stock price. You do some mental arithmetic and write down your order. The agent runs back to the market but there's so much competition, the price changes before the order gets put in. So the agent comes back.</p>
<p>And what do they ask?</p>
<p>The same question, but with different arguments. Here's your budget now, here's the price now. You plug them into the same mental function, and get a new order.</p>
<p>Sometimes it happens. Things change between when you calculate a thing and when the thing needs to happen. So you recalculate. But it's better to give that agent an order first in case it gets in and recalculate if it doesn't. That's called "optimistic".</p>
<p>What would happen if your function wasn't pure? It would get called many times and have more than one effect. It would print out something many times instead of one. Or it might launch two missiles instead of one. Oops. Keep it pure, folks.</p>
<p>Okay, there's just one more idea before we get to the catalog of primitives. It's an important idea, but I understand if you want to skip ahead. In fact, skip around as much as you want!</p>
<h2>Independence</h2>
<p>We talked before about being able to break up tasks into smaller tasks so you can get more concurrency and take advantage of more cores. How small do things need to be?</p>
<p>The short answer is "as small as possible". Why?</p>
<p>Let's say we have twenty people who have to carry 100 rocks of different sizes up a hill. They've each got a backpack. How do you distribute the rocks among the twenty backpacks so you can carry them all? It's like a game of Tetris. You could spend all day trying to figure out how to fit them all optimally. Most of the time, you end up with some rocks that just won't fit. So you move them around a little, trying to find some extra space somewhere. But it's a lot of work and most of the time you can't fit them all in. This is a well-known problem called the <a href="https://en.wikipedia.org/wiki/Knapsack_problem"><strong>Knapsack Problem</strong></a>.</p>
<p>What's the solution?</p>
<p>Break the rocks up. If you break them into gravel, you can literally pour them into the backpacks. And it's really easy. You don't have to think about it. But then you get to the end, and some still don't fit.</p>
<p>What's the solution?</p>
<p>Break them up even more! If you break them into sand, there's less empty space between them. And you can fill them up even more.</p>
<p>What does this have to do with concurrency?</p>
<p>It turns out that the same thing happens with tasks on your cores. If you have big tasks, you have to spend a lot of time figuring out which core would run which task to optimize the execution. But with tiny tasks, you don't think at all. You just pour them in wherever they fit. And more will fit! So break up your jobs into independent tasks and you will be able to run on more cores.</p>
<p>There's one caveat, though.</p>
<p>Sometimes you can break your task up into such small pieces, it's not worth the overhead of sending it to another core. For instance, adding two numbers is really fast on a computer. You might as well do it in the current thread instead of sending it to another machine. When you send it to another machine, you have to make a message, put it in a queue, and wait for the queue to process. It's not worth it! For many parallel execution examples, you'll see simple operations like addition used, just to make it easy to understand, even if you wouldn't really use that example in real life. However, in real-world calculations, you often have the opposite problem, which is that tasks are too big. And those could be broken up more.</p>
<p>So what does <strong>independent</strong> mean?</p>
<p>There are different types of independence. We've already looked at one type of independence. Pure functions are <strong>independent of time</strong>. It doesn't matter how many <strong>times</strong> you call them or <strong>when</strong> you call them. They always give the same answer. So that's one independence.</p>
<p>What's another?</p>
<p>Well, another one we could look at is <strong>order independence</strong>. For instance, if I need to fetch 100 different websites, it doesn't really matter what order I do them in. It <strong>does</strong> matter when I do the requests, because I could make the request when the server is down, or the page changes over time.</p>
<p>But the order doesn't matter. The answer from server A does not depend on the answer from server B. I could do them in any order.</p>
<p>And that's really important.</p>
<p>Let's look at an example. What if you're shopping at a grocery store with ten checkout lines? If you finish shopping before another person, does that mean you're going to finish checking out before them? No! If you're anything like me, you tend to choose the wrong line. That person who finished after you gets lucky and you watch them finish before you even started.</p>
<p>Ugh. It sucks.</p>
<p>But let's look at it analytically: each checkout is independent of the order. You still get the same answer. Everyone got their food. Everyone got paid. You got the same result. But what wasn't independent? You can't checkout before you shop!</p>
<p>Order independence is really strong. You can't always guarantee it, but when you can, you have a lot of options for breaking up that work. In this case, there are ten checkout centers, each with their own queue. The queues can move independently.</p>
<p>Mathematicians call order independence <strong>commutativity</strong>. Here's an example:</p>
<pre>
<code>a + b + c = b + a + c</code>
</pre>
<p>That is, addition is commutative. Why does commutativity matter? Because you can't always guarantee the order. If you divide up tasks among four cores, you can't guarantee they'll finish in the same order. If you need order, you won't be able to divide the tasks up.</p>
<p>Another type of independence?</p>
<p>Grouping. Grouping is a little hard to explain, so I'll need a nice example. I hope I won't disappoint.</p>
<p>Let's say I want to build the longest horizontal lego tower in the known universe. So to get started, I lay out all the legos in the order I want them. The color pattern, that is, the order is important. However, if I keep the order of the pieces, does it matter which ones I connect first? Not really. I'll get the same tower. Mathematicians call this "grouping". It's like in this equation:</p>
<pre>
<code>a + (b + c) = (a + b) + c</code>
</pre>
<p>The grouping is the parentheses. The letters are in the same order, but the operations happen differently. It's a subtle but important difference. Mathematicians call this <strong>associativity</strong>.</p>
<p>Why is this important?</p>
<p>Well, if I'm really going to make the longest lego tower in the known universe, I need to get some help. After laying out the legos, I can get some friends to help. All I have to do is to roughly divide up sections of legos among my friends, we can all connect up our legos (maintaining order!), then connect up the sections when we're done.</p>
<p>Because I can group them however I want and get the same answer, it lets me be flexible dividing up the work. I know I'll be able to put the pieces back together and get the same answer.</p>
<p>There are more types of independence, like <strong>idempotence</strong>. Idempotence means that if something happens twice (or more), it's the same as it happening once. So, for instance, if you add the same value to a set twice, it's the same as it happening once. Why is this important?</p>
<p>Well, here's another example. You're probably familiar with it.</p>
<p>Have you ever been on a site where you buy something and after you click "Buy" it yells at you:</p>
<p>"Don't click the buy button twice! It will charge you twice!"</p>
<p>Wouldn't the site be better with less yelling and more idempotence? Yes. The "submit payment" operation should be idempotent so you don't have to yell at the nice people who are trying to give you money. If you accidentally click twice, it should only charge you once.</p>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via PurelyFunctional.tv http://bit.ly/1UlCvdP<br>
<br>
April 10, 2019 at 10:42AM
